dashboard/test/lib/queue_processor_test.rb

```diff
@@ -0,0 +1,187 @@
+            @fail_on_message.delete(body)
+            raise "Fake failure for '#{body}'"
+          end
+        end
+        messages.each do |message|
+          @bodies.add(message.body)
+        end
+        @logger.debug "Handler succeeding on batch of size '#{messages.size}'"
+      end
+    end
+
+    # Fail the next handle attempt for the given message body.
+    def raise_on_next_receipt_of(body)
+      @lock.synchronize {
+        @logger.debug "Adding fake failure for #{body}"
+        @fail_on_message.add(body)
+      }
+    end
+
+    def received_bodies
+      @lock.synchronize {
+        @bodies.dup
+      }
+    end
+
+    def reset_received_bodies
+      @lock.synchronize {
+        @bodies = Set.new#<String>
+      }
+    end
+  end
+
+  def setup
+    # We need to allow http connections from the test to either the real or FAKE SQS service.
+    FakeWeb.allow_net_connect = true
+
+    @sqs = Aws::SQS::Client.new
+    unless ENV['USE_REAL_SQS'] == 'true'
+      $fake_sqs_service.start
+      @sqs.config.endpoint = $fake_sqs_service.uri
+    end
+
+    @logger = ENV['LOG_TO_STDOUT'] ? Logger.new(STDOUT) : Rails.logger
+    @lock = Mutex.new
+    @message_id = 0
+  end
+
+  def teardown
+    unless ENV['USE_REAL_SQS'] == 'true'
+      $fake_sqs_service.stop
+    end
+  end
+
+  def test_queue_processor
+    response = @sqs.create_queue(
+        queue_name: "test-queue-processor-test",
+        # Set a short visibility timeout so that retries will happen quickly.
+        attributes: {"VisibilityTimeout" => "1"}
+    )
+    queue_url = response.queue_url
+
+    # Set up a handler which will fail on the initial delivery attempt
+    # for 2 of the bodies so that we can make sure that those bodies are subsequently
+    # redelivered and handled.
+    handler = TestHandler.new(logger)
+
+    # Configure and start the queue processor.
+    sqs_metrics = SQS::Metrics.new
+    num_workers = 5
+    global_max_messages_per_sec = 25
+    config = SQS::QueueProcessorConfig.new(
+        queue_url: queue_url,
+        handler: handler,
+        initial_max_rate: global_max_messages_per_sec,
+        dcdo_max_rate_key: 'test-max-rate',
+        num_processors: 1,
+        num_workers_per_processor: num_workers,
+        logger: logger)
+    processor = SQS::QueueProcessor.new(config, sqs_metrics)
+    processor.start
+
+    # Enqueue 50 messages in batches of size 10.
+    batches = 5
+    batch_size = 10
+    num_messages = batches * batch_size
+    expected_bodies = Set.new#<String>
+    batches.times do
+      batch = []
+      batch_size.times do
+        item = {id: SecureRandom.uuid, message_body: create_unique_message_body}
+        expected_bodies << item[:message_body]
+        batch << item
+      end
+      @sqs.send_message_batch(queue_url: queue_url, entries: batch)
+    end
+
+    # Wait for the queue processor to handle all of the messages.
+    expected_elapsed_time = num_messages / global_max_messages_per_sec
```

Comment:
Lint/UselessAssignment: Useless assignment to variable - `expected_elapsed_time`.